sktime/forecasting/model_selection/tests/test_tune.py

```diff
@@ -353,35 +353,72 @@ def test_gscv_backends(backend_set):
     gscv.fit(y, X)
 
 
+def test_BaseGridSearch(
+    Forecaster,
+    *,
+    as_distribution=False,
+    **kwargs,
+):
+    if "param_distributions" not in kwargs and "param_grid" not in kwargs:
+        params = (
+            [
+                {
+                    "window_length": [1, 2, 3],
+                    "strategy": ["last", "mean"],
+                    "transformer__degree": [1, 2, 3],
+                    "forecaster__strategy": ["last", "mean", "seasonal_last"],
+                },
+                {
+                    "window_length": [4, 5, 6],
+                    "forecaster__strategy": ["last", "mean"],
+                },
+            ],
+        )
+        kwargs.setdefault(
+            "param_distributions" if as_distribution else "param_grid", params
+        )
+    return Forecaster, {
+        **kwargs,
+    }
+
+
+def test_ForecastingGridSearchCV(**kwargs):
+    return test_BaseGridSearch(ForecastingGridSearchCV, **kwargs)
+
+
+def test_ForecastingRandomizedSearchCV(**kwargs):
+    return test_BaseGridSearch(
+        ForecastingRandomizedSearchCV, as_distribution=True, **kwargs
+    )
+
+
+def test_ForecastingSkoptSearchCV(**kwargs):
+    return test_BaseGridSearch(ForecastingSkoptSearchCV, as_distribution=True, **kwargs)
+
+
 @pytest.mark.parametrize("return_n_best_forecasters", [-1, 0, 1, 2])
 @pytest.mark.parametrize(
-    "param_grid",
+    "Forecaster, kwargs",
     [
-        PIPE_GRID,
-        [
-            {
-                "forecaster__strategy": ["last", "mean"],
-                "transformer__forecaster__degree": [3, 4],
-            },
-            {
-                "forecaster__strategy": ["mean"],
-                "transformer__forecaster__degree": [5, 6, 7],
-            },
-        ],
+        test_ForecastingGridSearchCV(),
+        test_ForecastingRandomizedSearchCV(),
+        test_ForecastingRandomizedSearchCV(n_iter=3),
+        test_ForecastingRandomizedSearchCV(n_iter=5),
+        # test_ForecastingSkoptSearchCV(n_iter=5),  # Not working
     ],
 )
-def test_return_n_best_forecasters(return_n_best_forecasters, param_grid):
+def test_return_n_best_forecasters(Forecaster, return_n_best_forecasters, kwargs):
     y, X = load_longley()
-    gscv = ForecastingGridSearchCV(
+    gscv = Forecaster(
+        **kwargs,
         forecaster=PIPE,
-        param_grid=param_grid,
         cv=CVs[0],
         return_n_best_forecasters=return_n_best_forecasters,
     )
     gscv.fit(y, X)
     if return_n_best_forecasters == -1:
 
-        def calculate_total_combinations(param_grid: dict | list[dict]) -> int:
+        def calculate_total_combinations(param_grid):
```

Comment:
If you want, you could use Union and List. The | shirtcut is unavailable before 3.10, and list not before 3.9 I think.

Reply:
We need type ~script~ python

Reply:
Sorry, couldn't follow you. Can you please explain? 

I meant you can do this at start:

```python
from typing import List, Union
```

Reply:
> It was a joke. I was trying to be funny, but it didn't work sorry

I thought it was funny 😆

Reply:
You should be able to add `from __future__ import annotations` at the top of the file and use PEP 585 style annotations for earlier versions of Python (incl Python 3.8)

Reply:
Since sktime uses flake8, you could consider using this plugin: https://github.com/tyleryep/flake8-future-annotations

I don't personally use it but I use an [equivalent feature](https://docs.astral.sh/ruff/settings/#lint_isort_required-imports) with `ruff`.

Reply:
I already used `typing` module as instructed [here](https://github.com/sktime/sktime/pull/6031#discussion_r1509146401). Should I change it?

Reply:
Ultimately it's up to the maintainers. My take is that this should probably be tackled in a different PR if sktime decides to adopt PEP 585 (`List[int] -> list[int]`) and PEP 604 (`Union[str, int] -> str | int`) annotations. There are tools that could help automatically apply this migration to the entire codebase.

Reply:
the current problem with the mentioned PEPs is lack of 3.8 compatibility, and afaik a good number of users are still on 3.8. Though, 3.8 reaches EoL this year, after which we can in-principle add such annotations.

Reply:
The 2nd option will need 3.10, so one more year I suppose.

Or we can add typing-extensions as a dependency.

Reply:
@fkiraly @yarnabrina  the `from __future__ import annotations` import makes these compatible with Python 3.8. 

I just checked in two of our apps running in production w/ Python 3.8 and they both work with PEP 585 (`List[int] -> list[int]`) and PEP 604 (`Union[str, int] -> str | int`) style annotations when using in combination with `from __future__ import annotations`

Reply:
I understand, my comment was towards @fkiraly 's comment above.

I've nothing against future imports, but it doesn't enable all of the newer features. Addition of typing extensions will allow everything at a cost of very minimal, if any (it may already be a indirect dependency), dependency changes.

Reply:
I think it's better to err on the side of compatibility, future imports feel risky.

Apart from that - and this is not a strong opinion - I prefer to emphasize good dosctrings over type annotations, since:

* assumptions on input are typically stronger than just the type
* truthful type annotations are often hard to read
* docstrings are much lower entry barrier for new contributors

Example, from skbase, would you prefer

```python
    def _coerce_to_named_object_tuples(
        self,
        objs: Union[
            Sequence[Union[BaseObject, Tuple[str, BaseObject]]], Dict[str, BaseObject]
        ],
        clone: bool = False,
        make_unique: bool = True,
    ) -> List[Tuple[str, BaseObject]]:
```

or would you prefer reading the docstring?

```python
        """Coerce sequence of objects or named objects to list of (str, obj) tuples.

        Input that is sequence of objects, list of (str, obj) tuples or
        dict[str, object] will be coerced to list of (str, obj) tuples on return.

        Parameters
        ----------
        objs : list of objects, list of (str, object tuples) or dict[str, object]
            The input should be coerced to list of (str, object) tuples. Should
            be a sequence of objects, or follow named object API.
        clone : bool, default=False.
            Whether objects in the returned list of (str, object) tuples are
            cloned (True) or references (False).
        make_unique : bool, default=True
            Whether the str names in the returned list of (str, object) tuples
            should be coerced to unique str values (if str names in input
            are already unique they will not be changed).

        Returns
        -------
        list[tuple[str, BaseObject]]
            List of tuples following named object API.

            - If `objs` was already a list of (str, object) tuples then either the
              same named objects (as with other cases cloned versions are
              returned if ``clone=True``).
            - If `objs` was a dict[str, object] then the named objects are unpacked
              into a list of (str, object) tuples.
            - If `objs` was a list of objects then string names were generated based
               on the object's class names (with coercion to unique strings if
               necessary).
        """
```

Reply:
May be my personal preference, but both. They are not replacement for one another.

If you have tools like pylance or etc. installed, or even without them sometimes, type hints are very useful during development to get auto completions or suggestions that I am not passing objects of correct types. I'll of course prefer to use through a more meaningful type alias though.

But for first time use (not just just first, but more often at start of using a new package), just argument names or argument types may not be enough to get full idea what is an argument and then docstrings are much more helpful, when present. The possible issue with docstrings is that people (including me) often make a change in signature to try new things to do quick stuff, and then completely forget to update the docstring. Over time, docstrings and signature are barely aligned together, and since no linters (so far what I have used, may @tpvasconcelos can suggest something new?) check if docstrings note correctly as signature or not it's very difficult to detect.